cpp/include/cudf/detail/indexalator.cuh

```diff
@@ -0,0 +1,455 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_view.hpp>
+#include <cudf/utilities/traits.hpp>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief The base class for the input or output index normalizing iterator.
+ *
+ * This implementation uses CTRP to define the `input_indexalator` and the
+ * `output_indexalator` classes. This is so this class can manipulate the
+ * uniquely typed subclass member variable `p_` directly without requiring
+ * virtual functions since iterator instances will be copied to device memory.
+ *
+ * The base class mainly manages updating the `p_` member variable while the
+ * subclasses handle accessing individual elements in device memory.
+ *
+ * @tparam T The derived class type for the iterator.
+ */
+template <class T>
+struct base_indexalator {
+  using difference_type   = ptrdiff_t;
+  using value_type        = size_type;
+  using pointer           = size_type*;
+  using iterator_category = std::random_access_iterator_tag;
+
+  base_indexalator()                        = default;
+  base_indexalator(base_indexalator const&) = default;
+  base_indexalator(base_indexalator&&)      = default;
+  base_indexalator& operator=(base_indexalator const&) = default;
+  base_indexalator& operator=(base_indexalator&&) = default;
+
+  /**
+   * @brief Prefix increment operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T& operator++()
+  {
+    T& derived = static_cast<T&>(*this);
+    derived.p_ += width_;
+    return derived;
+  }
+
+  /**
+   * @brief Postfix increment operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T operator++(int)
+  {
+    T tmp{static_cast<T&>(*this)};
+    operator++();
+    return tmp;
+  }
+
+  /**
+   * @brief Prefix decrement operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T& operator--()
+  {
+    T& derived = static_cast<T&>(*this);
+    derived.p_ -= width_;
+    return derived;
+  }
+
+  /**
+   * @brief Postfix decrement operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T operator--(int)
+  {
+    T tmp(static_cast<T&>(*this));
```

Comment:
Sorry, I should have been clearer in my previous PR comment that there was multiple cases where `{}` could be used.
```suggestion
    T tmp{static_cast<T&>(*this)};
```

Reply:
Yes, that was my fault. I made these changes locally but pushed without them. Thanks.